### DISCUSSION TITLE
How Starlette Shapes UploadFile in FastAPI

### DISCUSSION THREAD

**Discussion by Kfir-G** [fastapi/fastapi#12296] [Show and tell]
Hey FastAPI community! 

I wanted to clarify something that often trips up developers using FastAPI: the `UploadFile` class and its relationship with Starlette. 

### What's the Confusion?

In FastAPI, the `UploadFile` class is key for handling file uploads. However, many developers find themselves confused when they check the type of an `UploadFile` instance. They discover it's actually an instance of Starlette’s `UploadFile`.

### Here’s the Lowdown:

1. **Architecture**: FastAPI is built on top of Starlette, which is a lightweight ASGI framework. When you use `UploadFile` in FastAPI, you're actually working with a subclass of Starlette’s `UploadFile`, which adds some handy features.

2. **Type Confusion**:
   Here's a quick example:
   ```python
   from fastapi import FastAPI, UploadFile, File

   app = FastAPI()

   @app.post("/upload/")
   async def upload_file(file: UploadFile = File(...)):
       return {
           "is_fastapi_instance": isinstance(file, UploadFile),
           "file_type": type(file),
           "filename": file.filename
       }
   ```
   In this case, `isinstance(file, UploadFile)` will return `False` ⚠️ , when you will check the type with `type(file)`, you’ll see it as `starlette.datastructures.UploadFile`.

3. **Why It Matters**: Understanding this inheritance is crucial for debugging and writing logic that relies on type checking. It highlights how FastAPI builds upon Starlette while adding its own capabilities.

### Additional Resources
If you're interested in digging deeper, check out these links:
- [FastAPI Documentation: Upload File](https://fastapi.tiangolo.com/tutorial/request-files/#uploadfile)
- [FastAPI UploadFile Class](https://github.com/fastapi/fastapi/blob/847296e885ed83bc333b6cc0a3000d6242083b87/fastapi/datastructures.py#L30)
- [Starlette Documentation: UploadFile](https://www.starlette.io/requests/#uploadfile)
- [Understanding ASGI](https://asgi.readthedocs.io/en/latest/)
- [Starlette UploadFile Class](https://github.com/encode/starlette/blob/fa7b382a66cd99e3dc18f3baa44dae5ec68be76b/starlette/datastructures.py#L409)
- [Understanding FastAPI: How Starlette Works](https://dev.to/ceb10n/understanding-fastapi-how-starlette-works-43i1)

I hope this helps clear things up.

**Comment by yangyushi** [comment]
Thanks for the clarification and this seems like a really confusing design choice to me. Would it be fixed at some point so that `fastapi.UploadFile` would work as one would expected naively at some point?
